CMakeLists.txt

```diff
@@ -62,99 +33,72 @@ include_directories (
     ${PROJECT_SOURCE_DIR}/rabit/include
 )
 
-file(GLOB SOURCES
-    src/c_api/*.cc
-    src/common/*.cc
-    src/data/*.cc
-    src/gbm/*.cc
-    src/metric/*.cc
-    src/objective/*.cc
-    src/tree/*.cc
+file(GLOB_RECURSE SOURCES
     src/*.cc
+    src/*.h
 )
 
+# TODO: Create rabit cmakelists.txt
 set(RABIT_SOURCES
     rabit/src/allreduce_base.cc
     rabit/src/allreduce_robust.cc
     rabit/src/engine.cc
     rabit/src/c_api.cc
 )
 
-file(GLOB CUDA_SOURCES
+file(GLOB_RECURSE CUDA_SOURCES
     plugin/updater_gpu/src/*.cu
-    plugin/updater_gpu/src/exact/*.cu
+    plugin/updater_gpu/src/*.cuh
 )
 
 add_subdirectory(dmlc-core)
-
-add_library(rabit STATIC ${RABIT_SOURCES})
-
-#Set library output directories
-if(MSVC)
-    #With MSVC shared library is considered runtime
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${PROJECT_SOURCE_DIR}/lib)
-    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELEASE ${PROJECT_SOURCE_DIR}/lib)
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG ${PROJECT_SOURCE_DIR}/lib)
-    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_DEBUG ${PROJECT_SOURCE_DIR}/lib)
-else()
-    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR}/lib)
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_SOURCE_DIR})
-endif()
-
 set(LINK_LIBRARIES dmlccore rabit)
 
+# GPU Plugin
 if(PLUGIN_UPDATER_GPU)
-  find_package(CUDA REQUIRED)
-  
-  # nccl
+  find_package(CUDA 7.5 REQUIRED)
+  cmake_minimum_required(VERSION 3.5)
   add_subdirectory(nccl)
-  set(NCCL_DIRECTORY ${PROJECT_SOURCE_DIR}/nccl)
-  include_directories(${NCCL_DIRECTORY}/src)
-
-  #Find cub
-  set(CUB_DIRECTORY ${PROJECT_SOURCE_DIR}/cub/)
-  include_directories(${CUB_DIRECTORY})
-
-  #Find googletest
-  set(GTEST_DIRECTORY "${CACHE_PREFIX}" CACHE PATH "Googletest directory")
-  include_directories(${GTEST_DIRECTORY}/include)
-
-  #gencode flags
-  set(GPU_COMPUTE_VER 35;50;52;60;61 CACHE STRING
-    "Space separated list of compute versions to be built against")
   
+  include_directories(
+    nccl/src
+    cub
+  )
+
   set(GENCODE_FLAGS "")
-  foreach(ver ${GPU_COMPUTE_VER})
-    set(GENCODE_FLAGS "${GENCODE_FLAGS}-gencode arch=compute_${ver},code=sm_${ver};")
-  endforeach()
+  format_gencode_flags("${GPU_COMPUTE_VER}" GENCODE_FLAGS)
   set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};--expt-extended-lambda;${GENCODE_FLAGS};-lineinfo;")
   if(NOT MSVC)
-    set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};-Xcompiler -fPIC")
+    set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS};-Xcompiler -fPIC; -std=c++11")
   endif()
 
   cuda_add_library(gpuxgboost ${CUDA_SOURCES} STATIC)
   target_link_libraries(gpuxgboost nccl)
   list(APPEND LINK_LIBRARIES gpuxgboost) 
   list(APPEND SOURCES plugin/updater_gpu/src/register_updater_gpu.cc) 
-else()
-  set(CUDA_OBJS "")
 endif()
 
+add_library(rabit STATIC ${RABIT_SOURCES})
```

Comment:
Ideally, we should allow the user to configure the backend, in the same way Makefiles do.